src/index.ts

```diff
@@ -1,73 +1,69 @@
+    return type.nullable ? `Option<${t}>` : t
   }
-}
 
-function jsonWrappedRustType(property: Property): string {
-  if (property.$ref) return `Json<types::${helpers.capitalize(property.$ref.name)}>`;
-  switch (property.type) {
-    case "string":
-      if (property.format === "date-time") {
-        return "Json<chrono::DateTime<chrono::Utc>>";
-      }
-      return "String";
-    case "number":
-      if (property.format === "float") {
-        return "Json<f32>";
-      }
-      if (property.format === "double") {
-        return "Json<f64>";
+  switch (type.kind) {
+    case 'string':
+      return optionalize('String')
+    case 'int32':
+      return optionalize('i32')
+    case 'float':
+      return optionalize('f32')
+    case 'double':
+      return optionalize('f64')
+    case 'byte':
+      return optionalize('byte')
+    case 'date-time':
+      return optionalize("chrono::DateTime<chrono::Utc>")
+    case 'boolean':
+      return optionalize('bool')
+    case 'array':
+      const arrayType = type as ArrayType
+      return optionalize(`Vec<${toRustTypeX(arrayType.elementType)}>`)
+    case 'buffer':
+      return optionalize('Vec<u8>')
+    case 'object':
+      const oType = (type as ObjectType)
+      if (oType.properties?.length > 0) {
+        return optionalize(`types::${helpers.capitalize(oType.name)}`)
+      } else {
+        // we're just exposing the serde values directly for backwards compat
+        return optionalize("std::collections::HashMap<String, serde_json::Value>")
       }
-      return "Json<i64>";
-    case "integer":
-      return "Json<i32>";
-    case "boolean":
-      return "Json<bool>";
-    case "object":
-      return "Json<std::collections::HashMap<String, serde_json::Value>>";
-    case "array":
-      if (!property.items) return "Json<Vec<serde_json::Value>>";
-      // TODO this is not quite right to force cast
-      return `Json<Vec<${toRustType(property.items as Property)}>>`;
-    case "buffer":
-      return "Vec<u8>";
+    case 'enum':
+      return optionalize(`types::${helpers.capitalize((type as EnumType).name)}`)
+    case 'map':
+      const { keyType, valueType } = type as MapType
+      return optionalize(`std::collections::HashMap<${toRustTypeX(keyType)}, ${toRustTypeX(valueType)}>`)
     default:
-      throw new Error("Can't convert property to Rust type: " + property.type);
+      throw new Error("Can't convert XTP type to Go type: " + type)
   }
 }
 
-function makePublic(s: string) {
-  return "pub " + s;
+function isOptional(type: String): boolean {
+  return type.startsWith('Option<')
+}
+
+function toRustType(property: XtpTyped, required?: boolean): string {
+  const t = toRustTypeX(property.xtpType)
+
+  // if required is unset, just return what we get back
+  if (required === undefined) return t
+
+  // if it's set and true, just return what we get back
+  if (required) return t
+
+  // otherwise it's false, wrap it in another Option
+  return `Option<${t}>`
```

Comment:
Hm. Double-wrapping `Option` is a little bit of an itchy sweater – is our goal to indicate to the plugin the status of the field _and_ the value? Or is it to emulate the behavior of other typed language PDKs on missing fields? (That is, missing required nullable fields should fail deserialization, and absent optional fields are not included in serialization?)

(By "optional" I mean "the field is not required", not that the value it refers to is wrapped with `Option`.)

My druthers would be to promote non-required fields to `Option` if they haven't already been promoted already by `nullable`, then to use `#[serde(default = "panic_if_empty")]` to enforce deserialization constraints. (That would look something [like this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7529f5c1ade67aeb5d08e1f36e57317e).)

For serialization we can check if the field is nullable + optional, and if so add a `#[serde(skip_deserializing_if="Option::is_none")]`.

Reply:
That would look something like the following (swapping `a_not_required` for `an_optional` in field names):

```rs
#[derive(
    Default, serde::Serialize, serde::Deserialize, extism_pdk::FromBytes, extism_pdk::ToBytes,
)]
#[encoding(Json)]
pub struct RequiredVsNullable {
    /// A not-required, not-nullable boolean prop
    #[serde(rename = "aNotRequiredNotNullableBoolean")]
    #[serde(default, skip_serializing_if="Option::is_none"))]
    pub an_optional_not_nullable_boolean: Option<bool>,

    /// A not-required, nullable boolean prop
    #[serde(rename = "aNotRequiredNullableBoolean")]
    #[serde(default, skip_serializing_if="Option::is_none")]
    pub an_optional_nullable_boolean: Option<bool>,

    /// A required, not-nullable boolean prop
    #[serde(rename = "aRequiredNotNullableBoolean")]
    pub a_required_not_nullable_boolean: bool,

    /// A required ,nullable boolean prop
    #[serde(rename = "aRequiredNullableBoolean")]
    #[serde(default = "fail_if_not_present")]
    pub a_required_nullable_boolean: Option<bool>,
}

fn fail_if_not_present() -> Option<bool> {
    panic!("Missing 'a_required_nullable_boolean' field in message");
}
```

Reply:
> (By "optional" I mean "the field is not required", not that the value it refers to is wrapped with Option.)

yes, i've taken to calling it "not required" as it's difficult to talk about.

> My druthers would be to promote non-required fields to Option if they haven't already been promoted already by nullable, then to use #[serde(default = "panic_if_empty")] to enforce deserialization constraints. (That would look something [like this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7529f5c1ade67aeb5d08e1f36e57317e).)

I'm fine with that. i felt reflecting the double option gives the user more control over what to do. but i could see how unwrapping it might be annoying.

Reply:
> My druthers would be to promote non-required fields to Option if they haven't already been promoted already by nullable, then to use #[serde(default = "panic_if_empty")] to enforce deserialization constraints. (That would look something [like this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7529f5c1ade67aeb5d08e1f36e57317e).)

okay, i'm having a hell of a time making this work generically 😆 in rust generics hell

```rust
fn panic_if_key_missing<'de, T, D>(deserializer: D) -> Result<T, D::Error>
where
    T: Deserialize<'de>,
    D: Deserializer<'de>,
{
    struct KeyMissingVisitor<T>(PhantomData<T>);

    impl<'de, T> Visitor<'de> for KeyMissingVisitor<T>
    where
        T: Deserialize<'de>,
    {
        type Value = T;

        fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
            formatter.write_str("a required key with any value or null")
        }

        fn visit_none<E>(self) -> Result<Self::Value, E>
        where
            E: DeError,
        {
            Err(E::custom("oh no: required key is missing"))
        }

        fn visit_some<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
        where
            D: Deserializer<'de>,
        {
            T::deserialize(deserializer)
        }
    }

    // Deserialize as `Option<T>` to detect if the key is missing
    let option = deserializer.deserialize_option(KeyMissingVisitor::<Option<T>>(PhantomData))?;
    
    option.ok_or_else(|| D::Error::custom("oh no: required key is missing"))
}
```

Reply:
Oh! you don't have to deal with all that, I don't think -- the function path provided to `#[serde(default = "path")]` only fires if the key is missing from the input during deserialization, so you should just be able to have a plain function that panics? (double check me on that though-- you're probably seeing something I'm not!)

Reply:
it seems to want the signature of `default` to match the property:

```

    ^^^^^^^^^^^^^^^^^ expected `Option<bool>`, found `()`
|
    = note:   expected enum `std::option::Option<bool>`
            found unit type `()`
```

Reply:
A generic might work there:

```rust
fn panic_if_not_given<T>() -> T {
    panic!("oh no");
}
```

Or if that doesn't work:

```rust
fn panic_if_not_given() -> ! {
    panic!("oh no");
}
```

Reply:
thanks! yes using never was the way